pkg/cmd/certregenerationcontroller/csrapprover.go

```diff
@@ -0,0 +1,295 @@
+	}
+	// Don't recreate readyControlPlaneNodesOnce as we want to keep nodes which are now not ready (and thus not present in this map)
+	if len(c.readyControlPlaneNodesOnce) == 0 {
+		c.readyControlPlaneNodesOnce = make(map[string]NodeAddresses, 0)
+	}
+	for _, node := range nodeList.Items {
+		if !checkNodeReady(&node) {
+			continue
+		}
+		dnsAddresses, internalIPAddresses := getNodeAddresses(&node)
+		c.readyControlPlaneNodesOnce[node.Name] = NodeAddresses{
+			DNSNames:    dnsAddresses,
+			IPAddresses: internalIPAddresses,
+		}
+	}
+	return nil
+}
+
+func (c *controller) getUsernames() sets.String {
+	userNames := sets.String{}
+	for node := range c.readyControlPlaneNodesOnce {
+		userNames = userNames.Insert(fmt.Sprintf("system:node:%s", node))
+	}
+	return userNames
+}
+
+func (c *controller) syncCSR(ctx context.Context, syncCtx factory.SyncContext) error {
+	klog.Infof("ControlPlaneNodeApprover: sync queue, key: %v", syncCtx.QueueKey())
+	csr, err := c.csrLister.Get(syncCtx.QueueKey())
+	if err != nil {
+		klog.Infof("ControlPlaneNodeApprover: error getting CSR from the queue: %w", err)
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	klog.Infof("ControlPlaneNodeApprover: syncing CSR %s", csr.Name)
+
+	csrCopy := csr.DeepCopy()
+	csrPEM, _ := pem.Decode(csr.Spec.Request)
+	if csrPEM == nil {
+		return fmt.Errorf("failed to PEM-parse the CSR block in .spec.request: no CSRs were found")
+	}
+
+	approved, denied := getCertApprovalCondition(&csrCopy.Status)
+	if approved {
+		return fmt.Errorf("CSR %s ignored: already approved", csrCopy.Name)
+	}
+	if denied {
+		return fmt.Errorf("CSR %s ignored: already denied", csrCopy.Name)
+	}
+
+	x509CSR, err := x509.ParseCertificateRequest(csrPEM.Bytes)
+	if err != nil {
+		return fmt.Errorf("failed to parse the CSR bytes: %v", err)
+	}
+
+	if err = c.Validate(csr, x509CSR); err != nil {
+		klog.Warningf("CSR %s ignored: %v", csrCopy.Name, err)
+		return err
+	}
+
+	klog.Infof("ControlPlaneNodeApprover: Approving CSR %s", csrCopy.Name)
+	csrCopy.Status.Conditions = append(csrCopy.Status.Conditions,
+		certapiv1.CertificateSigningRequestCondition{
+			Type:    certapiv1.CertificateApproved,
+			Status:  corev1.ConditionTrue,
+			Reason:  "AutoApproved",
+			Message: fmt.Sprintf("Auto-approved CSR %q", csrCopy.Name),
+		})
+	syncCtx.Recorder().Eventf("CSRApproval", "The CSR %q has been approved", csrCopy.Name)
+	_, err = c.csrClient.UpdateApproval(ctx, csrCopy.Name, csrCopy, metav1.UpdateOptions{})
+	return err
+}
+
+func (c *controller) Validate(csrObj *certapiv1.CertificateSigningRequest, x509CSR *x509.CertificateRequest) error {
+	if csrObj == nil || x509CSR == nil {
+		return fmt.Errorf("received a 'nil' CSR")
+	}
+
+	klog.Infof("ControlPlaneNodeApprover: working on CSR %s", csrObj.Name)
+	// Check username. It should either be a node bootstrapper or match one of known nodes
+	if csrObj.Spec.Username != bootstrapperUsername && !c.getUsernames().Has(csrObj.Spec.Username) {
```

Comment:
During my investigation on the cert-recovery story I found out that the `bootstrapperUsername` is a public asset. In theory any process running on a node can access it and issue a CSR. So, it looks like we cannot trust this asset and it looks like we need a different solution.

Reply:
Correct, there is no way for us to determine that this CSR genuinely comes from the node. However, we ensure that the approved cert will be valid for node IPs and DNSes only, it would be useless for any other node, so its not that dangerous to approve it

Reply:
> However, we ensure that the approved cert will be valid for node IPs and DNSes only, it would be useless for any other node, so its not that dangerous to approve it

I think that kube doesn't validate if a client certificate comes from a specific IP address. Anyone with the certificate can claim the identity.

Reply:
Correct, but this identity is usable on the particular master node. Also, this CSR will be approved only for refreshes, so a new node certs won't be approved. I can't imagine a way this cert to be useful for the attacker

Reply:
> Also, this CSR will be approved only for refreshes

How do you know the old certificate has expired ?

> Correct, but this identity is usable on the particular master node

How can you ensure that since there is no way to check on the server if a request comes from a specific machine ?

>  I can't imagine a way this cert to be useful for the attacker

If the signer for a CSR is public then any process on any machine can read it and send a CSR request, right ?

Reply:
> The certificate has a set of IPs and DNSNames its valid for. It can't be used from any other node

If any process can generate a certificate with exactly the same content (IPs, DNSNames) then how can we ensure anything ?

Note that in cryptography it doesn't matter which IP sent a request. What matters is presenting a valid secret, in this case, a certificate.

Can you point me to code that check if the IP address of a sender is associated with a certificate ?